src/librustc/mir/mod.rs

```diff
@@ -1779,31 +1785,100 @@ pub type PlaceElem<'tcx> = ProjectionElem<'tcx, Local, Ty<'tcx>>;
 
 newtype_index!(Field { DEBUG_FORMAT = "field[{}]" });
 
-impl<'tcx> Place<'tcx> {
-    pub fn field(self, f: Field, ty: Ty<'tcx>) -> Place<'tcx> {
-        self.elem(ProjectionElem::Field(f, ty))
+impl<'a, 'tcx> Place<'tcx> {
+    // projection lives in the last elem.
+    pub fn projection(&self) -> Option<&PlaceElem> {
+        self.elems.last()
     }
 
-    pub fn deref(self) -> Place<'tcx> {
-        self.elem(ProjectionElem::Deref)
+    pub fn field(
+        self,
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        f: Field,
+        ty: Ty<'tcx>,
+    ) -> Self {
+        self.elem(tcx, ProjectionElem::Field(f, ty))
     }
 
-    pub fn downcast(self, adt_def: &'tcx AdtDef, variant_index: usize) -> Place<'tcx> {
-        self.elem(ProjectionElem::Downcast(adt_def, variant_index))
+    pub fn deref(self, tcx: TyCtxt<'a, 'tcx, 'tcx>) -> Self {
+        self.elem(tcx, ProjectionElem::Deref)
     }
 
-    pub fn index(self, index: Local) -> Place<'tcx> {
-        self.elem(ProjectionElem::Index(index))
+    pub fn downcast(
+        self,
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        adt_def: &'tcx AdtDef,
+        variant_index: usize,
+    ) -> Self {
+        self.elem(tcx, ProjectionElem::Downcast(adt_def, variant_index))
     }
 
-    pub fn elem(self, elem: PlaceElem<'tcx>) -> Place<'tcx> {
-        Place::Projection(Box::new(PlaceProjection { base: self, elem }))
+    pub fn index(self, tcx: TyCtxt<'a, 'tcx, 'tcx>, index: Local) -> Self {
+        self.elem(tcx, ProjectionElem::Index(index))
+    }
+
+    pub fn constant_index(
+        self,
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        offset: u32,
+        min_length: u32,
+        from_end: bool,
+    ) -> Self {
+       self.elem(tcx, ProjectionElem::ConstantIndex {
+           offset, min_length, from_end,
+       })
+    }
+
+    pub fn subslice(
+        self,
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        from: u32,
+        to: u32,
+    ) -> Self {
+        self.elem(tcx, ProjectionElem::Subslice {
+            from, to,
+        })
+    }
+
+    pub fn elem(
+        self,
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        elem: PlaceElem<'tcx>,
+    ) -> Self {
+        Place {
+            base: self.base,
+            elems: tcx.intern_place_elems(&[elem]),
```

Comment:
This replaces the current `elems` - it should instead have `elem` at the end.

Reply:
**Minor** nit.

The builder pattern used here suggests that the `Place` should be built up incrementally, but it seems to me that it would unnecessarily intern every single intermediate projection. That is, the code currently does not expose any way to construct the whole projection in one go, so one would end up doing something along the lines of

```
let place = Place::local(x);
let place = place.such_projection(...);
let place = place.another_projection(...);
...
```

and thus having every slice for each intermediate projection interned. Not sure if this is a relevant problem or whether the current code would be able to take advantage of a one-shot method, though.